### PR TITLE
add ISOType 

### LINKED ISSUE #11: typeFromIso
Not sur about the naming, would something like that be ok?

```typescript
export const typeFromIso = <I, O>(iso: Iso<I, O>, is: (x: O) => x is O,  name: string): t.Type<I, O> =>
  new t.Type<I, O>(name, is, (s, _) => t.success(iso.to(s)), iso.from)
```

### PR DESCRIPTION
**Pull request by sledorze**
Fixes #11

### CODE REVIEW

File: src/monocle-ts/ISOType.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import { Iso } from 'monocle-ts'
+import * as t from 'io-ts'
+
+export const ISOType = <I, O>(name: string, iso: Iso<I, O>, is: (x: O) => x is O): t.Type<I, O> =>
```
Comment (gcanti) [inline src/monocle-ts/ISOType.ts]: Can we name this function `typeFromIso` as you proposed, or even better just `fromIso`?
The signature of `is` should be `is: (x: any) => x is O`...

File: test/index.ts [hunk 2]
```diff
@@ -129,6 +131,28 @@ describe('monocle-ts', () => {
...
+      a: string
+    }
+    const a: A = { a: 'x' }
+    type B = {
+      b: string
+    }
+    const b: B = { b: 'x' }
+
+    const ABiso = new m.Iso<A, B>(n => ({ b: n.a }), n => ({ a: n.b }))
+    const ABType = ISOType('t', ABiso, (n): n is B => n.hasOwnProperty('b') && typeof n['b'] === 'string')
```
Comment (gcanti) [inline test/index.ts]: ...so here the `is` function should check that `n` is an object in the first place

**Comment by gcanti** [comment]
Based on the test I'm not sure what's the use case for this new function though, could you please elaborate?

**Review by gcanti — CHANGES_REQUESTED** [review]

File: src/monocle-ts/ISOType.ts (hunk 1 above)
Comment (sledorze) [reply src/monocle-ts/ISOType.ts]: oh yes, my bad

File: test/index.ts (hunk 2 above)
Comment (sledorze) [reply test/index.ts]: indeed

**Comment by sledorze** [comment]
@gcanti I have created some newTypes (from  https://github.com/gcanti/newtype-ts) which are Iso, and want to quickly be able to put some Type definition.

**Comment by gcanti** [comment]
Do you want to keep the `ISOType.ts` filename or it's an oversight?

> want to quickly be able to put some Type definition

Ok but what would you do with the resulting type?

```ts
import { Newtype, iso } from 'newtype-ts'

type Age = Newtype<'Age', number>
const ageIso = iso<Age>()

import { Iso } from 'monocle-ts'
import * as t from 'io-ts'

export const fromIso = <I, O>(name: string, iso: Iso<I, O>, is: (x: any) => x is O): t.Type<I, O> =>
  new t.Type<I, O>(name, is, (s, _) => t.success(iso.to(s)), iso.from)

// t.Type<Newtype<"Age", number>, number>
const T = fromIso('T', ageIso, t.number.is)

// ???
```

**Comment by sledorze** [comment]
@gcanti I would do that:
Put it as part of a `t.interface({ age: T })`BUT before being able to do that I would first reverse the Iso (otherwise the conversion is in the wrong order).

Hence this former request on monocle-ts:
https://github.com/gcanti/monocle-ts/issues/36

I hope it make sense.

as for the Naming, rereading the content of the repo I understand that for homogeneity as I export only one function I should make the function and file type the same.
I'll push an update now.

**Comment by gcanti** [comment]
> Put it as part of a t.interface({ age: T })

It would be unsafe since `T: t.Type<number, Age>` (should be `T: t.Type<any, Age>`)

```ts
// T: t.Type<number, Age>
const T = fromIso('T', ageIso.reverse(), (v): v is Age => t.number.is(v))

const P = t.interface({ age: T })

console.log(t.validate({ age: true }, P)) // right(...)
```

Wouldn't a `fromNewtype` function be more handy?

```ts
import { AnyNewtype, Carrier } from 'newtype-ts'

const fromNewtype = <N extends AnyNewtype>(type: t.Type<any, Carrier<N>>): t.Type<any, N> => type as any

const T = fromNewtype<Age>(t.number)

const P = t.interface({ age: T })

console.log(t.validate({ age: true }, P)) // left(...)
```

**Comment by sledorze** [comment]
@gcanti true, it would requier to be piped after a `t.number`.
`fromNewType` may be handy yes, and would fill my need, I was just thinking more broadly but when I try to find other examples most of them are Prisms and not Isos.
I'll change for your suggestion.

**Review by gcanti — APPROVED** [review]

**Comment by gcanti** [comment]
added some docs https://github.com/gcanti/io-ts-types#fromnewtype

**Pull request merged**